DependencyInjection/Compiler/EntityListenerPass.php

```diff
@@ -41,35 +45,44 @@ public function process(ContainerBuilder $container)
                 }
 
-                if (isset($attributes['lazy']) && $attributes['lazy']) {
+                $class = $resolver->getClass();
+
+                if (substr($class, 0, 1) === '%') {
+                    // resolve container parameter first
+                    $class = $container->getParameterBag()->resolveValue($resolver->getClass());
+                }
+                $resolverSupportsLazyListeners = is_a($class, EntityListenerServiceResolver::class, true);
+
+                $lazyByAttribute = isset($attributes['lazy']) && $attributes['lazy'];
+                if ($lazyByAttribute && ! $resolverSupportsLazyListeners) {
+                    throw new InvalidArgumentException(sprintf(
+                        'Lazy-loaded entity listeners can only be resolved by a resolver implementing %s.',
+                        EntityListenerServiceResolver::class
+                    ));
+                }
+
+                if (! isset($attributes['lazy']) && $resolverSupportsLazyListeners || $lazyByAttribute) {
                     $listener = $container->findDefinition($id);
 
                     if ($listener->isAbstract()) {
                         throw new InvalidArgumentException(sprintf('The service "%s" must not be abstract as this entity listener is lazy-loaded.', $id));
                     }
 
-                    $interface = 'Doctrine\\Bundle\\DoctrineBundle\\Mapping\\EntityListenerServiceResolver';
-                    $class     = $resolver->getClass();
-
-                    if (substr($class, 0, 1) === '%') {
-                        // resolve container parameter first
-                        $class = $container->getParameterBag()->resolveValue($resolver->getClass());
-                    }
+                    $resolver->addMethodCall('registerService', [$listener->getClass(), $id]);
 
-                    if (! is_a($class, $interface, true)) {
-                        throw new InvalidArgumentException(
-                            sprintf('Lazy-loaded entity listeners can only be resolved by a resolver implementing %s.', $interface)
-                        );
+                    if (! isset($lazyServiceReferencesByResolver[$resolverId])) {
+                        $lazyServiceReferencesByResolver[$resolverId] = [];
                     }
-
-                    $listener->setPublic(true);
-
-                    $resolver->addMethodCall('registerService', [$listener->getClass(), $id]);
+                    $lazyServiceReferencesByResolver[$resolverId][$id] = new Reference($id);
                 } else {
                     $resolver->addMethodCall('register', [new Reference($id)]);
                 }
             }
         }
+
+        foreach ($lazyServiceReferencesByResolver as $resolverId => $listenerReferences) {
+            $container->findDefinition($resolverId)->replaceArgument(0, ServiceLocatorTagPass::register($container, $listenerReferences));
```

Comment:
@alcaeus there is one possible BC break with this I just thought about.

What if someone uses a custom entity listener resolver that implements `EntityListenerServiceResolver` but is not an instance of `ContainerEntityListenerResolver`?

Then this might fail as we cannot guarantee that the first constructor argument is actually a `ContainerInterface`...

So should we maybe only do this optimization if its our class `ContainerEntityListenerResolver`?

Reply:
I fixed this case. So now we are only doing this in case the class is `ContainerEntityListenerResolver`